partitura/musicanalysis/note_array_to_score.py

```diff
@@ -0,0 +1,275 @@
+        warnings.warn("Inferring measures", stacklevel=2)
+        score.add_measures(part)
+
+        warnings.warn("Find and tie notes", stacklevel=2)
+        # tie notes where necessary (across measure boundaries, and within measures
+        # notes with compound duration)
+        score.tie_notes(part)
+
+        warnings.warn("find and ensure tuplets", stacklevel=2)
+        # apply simplistic tuplet finding heuristic
+        score.find_tuplets(part)
+
+    warnings.warn("done create_part", stacklevel=2)
+    return part
+
+
+def note_array_to_score(note_array: Union[np.ndarray, list], part_id="", divs: int = None, assign_note_ids: bool = True,
+                       ensurelist: bool = False, estimate_key: bool = False, sanitize: bool = True, scorify=True) -> ScoreLike:
+    """
+    A generic function to transform an enriched note_array to part.
+
+    The function can be used for many different occasions, i.e. load_score_midi, part_from_match, part_from_graph, etc.
+    This function requires a note array that contains time signatures and key signatures(optional - can also estimate it automatically).
+    Note array should contain the following fields:
+    - onset_div or onset_beat
+    - duration_div or duration_beat
+    - pitch
+    - ts_beats
+    - ts_beat_type
+    - key_mode(optional)
+    - key_fifths(optional)
+    - id(required but can also be empty)
+
+    Parameters
+    ----------
+    note_array : structure array or list of structured arrays
+    divs : int (optional)
+        Necessary Provided divs for midi import.
+    assign_note_ids : bool (optional)
+        Assign note_ids
+    ensurelist: bool (optional)
+        ensure that output part is a list
+    estimate_key: bool (optional)
+
+    Returns
+    -------
+    part : list or Part or PartGroup
+        Maybe should return score
+    """
+    if isinstance(note_array, list):
+        parts = [
+            note_array_to_score(note_array=x, part_id=str(i), assign_note_ids=assign_note_ids, ensurelist=ensurelist, scorify=False) for
+            i, x in enumerate(note_array)]
+        return score.Score(partlist=parts)
+
+    if not isinstance(note_array, np.ndarray):
+        raise TypeError("The note array does not have the correct format.")
+    if len(note_array) == 0:
+        raise ValueError("The note array is empty.")
+
+    if "id" not in note_array.dtype.names:
+        note_ids = ["{}n{:4d}".format(part_id, i) for i in range(len(note_array))]
+        note_array = rfn.append_fields(note_array, "id", np.array(note_ids, dtype='<U256'))
+    elif assign_note_ids or np.all(note_array["id"] == note_array["id"][0]):
+        note_ids = ["{}n{:4d}".format(part_id, i) for i in range(len(note_array))]
+        note_array["id"] = np.array(note_ids)
+
+    dtypes = note_array.dtype.names
+    # check if note array contains time signatures
+    if not "ts_beats" in dtypes:
+        raise AttributeError("The note array does not contain a time signature.")
```

Comment:
Should the time signature information in the note array be required for this method? As we discussed last Friday, we could alternative just give an iterable with time signatures and times, and use that to create the time signatures instead (or if no time signature is added, assume a default time signature).

Reply:
I thought it was a great idea at the time, but needed some brainstorming on deciding how to do it. Currently, you can input a note array in 3 different forms, we have to adapt everything to work with all three (i.e. beats, divs, or beats normalized by bar and global time). I left this open for future implementation.

Reply:
Ok, we should think how to properly include this in the next version.